haystack/nodes/preprocessor/splitter.py

```diff
@@ -0,0 +1,755 @@
+
+                            If the number of units is irrelevant, `split_length` can be safely set at 0.
+
+        :param add_page_number: Add the number of the page a paragraph occurs in to the Document's meta
+                                field `"page"`. Page boundaries are determined by `"\f"' character which is added
+                                in between pages by `PDFToTextConverter`, `TikaConverter`, `ParsrConverter` and
+                                `AzureConverter`.
+        """
+        split_by = split_by if split_by is not None else self.split_by
+        split_regex = split_regex if split_regex is not None else self.split_regex
+        split_length = split_length if split_length is not None else self.merger.window_size
+        split_overlap = split_overlap if split_overlap is not None else self.merger.window_overlap
+        max_chars = max_chars if max_chars is not None else self.max_chars
+        max_tokens = max_tokens if max_tokens is not None else self.max_tokens
+        add_page_number = add_page_number if add_page_number is not None else self.add_page_number
+
+        self._validate_split_params(
+            split_by=split_by,
+            split_regex=split_regex,
+            split_length=split_length,
+            split_overlap=split_overlap,
+            max_chars=max_chars,
+            max_tokens=max_tokens,
+        )
+
+        if any(document.content_type != "text" for document in documents):
+            raise ValueError(
+                "Some documents do not contain text. Make sure to pass only text documents to this node. "
+                "You can use a RouteDocuments node to make sure only text documents are sent to the DocumentCleaner."
+            )
+
+        if split_by == "token" and not self.tokenizer:
+            raise ValueError(
+                "If you set split_by='token', you must give a value to 'tokenizer_model'. "
+                "Use the same model you're using for your Reader."
+            )
+
+        if split_by == "character":
+            splitter_function = lambda text: (text, [0] * len(text)) if text != "" else ([""], [0])
+
+        elif split_by == "regex":
+            if not split_regex:
+                raise ValueError("If 'split_by' is set to 'regex', you must give a value to 'split_regex'.")
+            splitter_function = partial(self.split_by_regex, pattern=split_regex)
+
+        elif split_by == "page":
+            splitter_function = lambda text: self.split_by_regex(text=text, pattern="\f")
+
+        elif split_by == "paragraph":
+            splitter_function = lambda text: self.split_by_regex(text=text, pattern="\n\n")
+
+        elif split_by == "sentence":
+            splitter_function = lambda text: self.split_by_sentence_tokenizer(text=text)
+
+        elif split_by == "word":
+            splitter_function = lambda text: self.split_by_regex(text=text, pattern="\s+")
```

Comment:
I think this will make the tokenization slower because the regex in `split_by_regex` is compiled for every document. Compiling it once and then re-using it for every document should be faster.

Reply:
I think I would pre-compile all known regexes (page, paragraph, word) on instantiation of the splitter, you could then have explicit functions for each of those and also would not need these lambdas.

For custom regexes I'd also pre-compile on instantiation if provided but keep open the option to pass it in when running the splitter. In that case, I would still pass the compiled regex to the `split_by_regex` function so that it doesn't have to be re-compiled every time.